### PR TITLE
Add slide in autocompletion (closes #791)

### LINKED ISSUE #791: (🎁) Slide in should have auto complete for easier picking from existing branches
image
image

Awesome plugin, can't believe I haven't been using it.

### PR DESCRIPTION
**Pull request by mkondratek**
depends on #957

### CODE REVIEW

**Comment by PawelLipski** [comment]
In the meantime... I've got this version installed locally, will use it for a while. Can't see any issues so far :)

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/SlideInDialog.kt [hunk 1]
```diff
@@ -1,129 +1,272 @@
...
-        object : DocumentAdapter() {
-          override fun textChanged(e: DocumentEvent) {
-            startTrackingValidation()
-            component.document.removeDocumentListener(this)
-          }
-        }
+  private fun updateBranchesField() {
+    val branches =
+      GitBranchUtil.sortBranchNames(gitRepository.branches.localBranches.map { it.name })
+        .filter { it != parentName }
```
Comment (mkondratek) [inline frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/SlideInDialog.kt:199]: Fix for https://github.com/VirtusLab/git-machete-intellij-plugin/issues/977. I am closing the issue. I ve got one thing to be resolved here so please **hold on with the further review for now**.

File: src/test/java/com/virtuslab/archunit/UiThreadUnsafeMethodInvocationsTestSuite.java [hunk 2]
```diff
@@ -109,11 +109,18 @@ public void check(JavaMethod method, ConditionEvents events) {
                     "git4idea.GitUtil.getRepositories(com.intellij.openapi.project.Project)",
                     "git4idea.GitUtil.getRepositoryManager(com.intellij.openapi.project.Project)",
                     "git4idea.branch.GitBranchesCollection.findLocalBranch(java.lang.String)",
+                    "git4idea.branch.GitBranchesCollection.getLocalBranches()",
```
[the remark this answers is not in the document]
Comment (mkondratek) [reply src/test/java/com/virtuslab/archunit/UiThreadUnsafeMethodInvocationsTestSuite.java:112]: you were right. I tihnk only the `runCommand` should not be called on UI thread however I managed to remove a few methods more - all reviewed

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/SlideInDialog.kt (hunk 1 above)
Comment (mkondratek) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/SlideInDialog.kt:201]: ready for rereview

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/SlideInDialog.kt [hunk 3]
```diff
@@ -1,129 +1,274 @@
...
+   * |  remotes/origin/HEAD -> origin/master
+   * ```
+   */
+  @UIThreadUnsafe
+  private fun loadUnmergedBranchesForRoot(root: VirtualFile): List<String>? {
+    var result: List<String>? = null
+
+    val handler =
+      GitLineHandler(project, root, GitCommand.BRANCH).apply {
+        addParameters("--no-color", "-a", "--no-merged")
```
Comment (PawelLipski) [inline frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/SlideInDialog.kt:130]: Hmmm `git branch` is still a **porcelain** command... I'd rather resort to the underlying **plumbing** command, `git for-each-ref`. See `git machete status -v` for an example invocation

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/SlideInDialog.kt (hunk 1 above)
Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/SlideInDialog.kt:201]: Lemme install it locally, as per _tradition_

File: src/test/java/com/virtuslab/archunit/UiThreadUnsafeMethodInvocationsTestSuite.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (PawelLipski) [reply src/test/java/com/virtuslab/archunit/UiThreadUnsafeMethodInvocationsTestSuite.java:112]: Okay... so in particular, `getLocalBranches()` (as it's the only one that looks suspicious from the newly-added ones) just reads branches from memory?

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/SlideInDialog.kt [hunk 4]
```diff
@@ -1,129 +1,274 @@
...
+  private fun loadUnmergedBranchesInBackground() =
+    ProgressManager.getInstance()
+      .run(
+        object :
+          Task.Backgroundable(
+            project,
+            getString(
+              "action.GitMachete.BaseSlideInBelowAction.dialog.slide-in.progress"
+            ),
+            true
```
Comment (PawelLipski) [inline frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/SlideInDialog.kt:101]: Nit: name the param

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/SlideInDialog.kt [hunk 5]
```diff
@@ -1,129 +1,274 @@
...
+  private fun createBranchField() =
+    ComboBoxWithAutoCompletion(MutableCollectionComboBoxModel(mutableListOf<String>()), project)
+      .apply {
+        prototypeDisplayValue = GIT_REF_PROTOTYPE_VALUE
+        setPlaceholder(
+          getString(
+            "action.GitMachete.BaseSlideInBelowAction.dialog.slide-in.placeholder"
+          )
+        )
+        setUI(DarculaComboBoxUI(0f, Insets(1, 0, 1, 0), false))
```
Comment (PawelLipski) [inline frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/SlideInDialog.kt:262]: Pls name `o0f` and `false`

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/SlideInDialog.kt (hunk 3 above)
Comment (mkondratek) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/SlideInDialog.kt]: actually `unmergedBranches` is not used so the whole loading is redundant

Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/SlideInDialog.kt]: Wow okay... why was it used originally but isn't anymore? 🤔

Comment (mkondratek) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/SlideInDialog.kt]: This class is based on the other IntelliJ dialog. I was removing some redundant parts and tweaking the others - left accidentally

**Comment by PawelLipski** [comment]
YOLO, given #1084, let's release it (or as for now, at least merge the entire chain to `develop`)

**Review by PawelLipski — APPROVED** [review]

File: src/test/java/com/virtuslab/archunit/UiThreadUnsafeMethodInvocationsTestSuite.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mkondratek) [reply src/test/java/com/virtuslab/archunit/UiThreadUnsafeMethodInvocationsTestSuite.java:112]: It reads from memory.

```
public GitRepositoryImpl::GitBranchesCollection getBranches() {
    GitRepoInfo info = myInfo;
    return new GitBranchesCollection(info.getLocalBranchesWithHashes(), info.getRemoteBranchesWithHashes());
  }

// and then 

  public Collection<GitLocalBranch> GitBranchesCollection::getLocalBranches() {
    return Collections.unmodifiableCollection(myLocalBranches.keySet());
  }
```

**Pull request merged**